Sources/SwiftProtobuf/SwiftProtobufError.swift

```diff
@@ -0,0 +1,620 @@
+                    return "Error unpacking a Google_Protobuf_Any message"
+                case .protoFileToModuleMappingError:
+                    return "Error with proto file to module mapping"
+                }
+            }
+        }
+
+        public var description: String {
+            String(describing: self.code)
+        }
+
+        private var code: Wrapped
+        private init(_ code: Wrapped) {
+            self.code = code
+        }
+
+        public static var binaryEncodingError: Self {
+            Self(.binaryEncodingError)
+        }
+        
+        public static var binaryDecodingError: Self {
+            Self(.binaryDecodingError)
+        }
+        
+        public static var jsonEncodingError: Self {
+            Self(.jsonEncodingError)
+        }
+        
+        public static var jsonDecodingError: Self {
+            Self(.jsonDecodingError)
+        }
+        
+        public static var textFormatDecodingError: Self {
+            Self(.textFormatDecodingError)
+        }
+        
+        public static var anyUnpackError: Self {
+            Self(.anyUnpackError)
+        }
+        
+        public static var protoFileToModuleMappingError: Self {
+            Self(.protoFileToModuleMappingError)
+        }
+    }
+
+    /// A location within source code.
+    public struct SourceLocation: Sendable, Hashable {
+        /// The function in which the error was thrown.
+        public var function: String
+
+        /// The file in which the error was thrown.
+        public var file: String
+
+        /// The line on which the error was thrown.
+        public var line: Int
+
+        public init(function: String, file: String, line: Int) {
+            self.function = function
+            self.file = file
+            self.line = line
+        }
+
+        @usableFromInline
+        internal static func here(
+            function: String = #function,
+            file: String = #fileID,
+            line: Int = #line
+        ) -> Self {
+            return SourceLocation(function: function, file: file, line: line)
+        }
+    }
+}
+
+extension SwiftProtobufError: CustomStringConvertible {
+    public var description: String {
+        "\(self.code): \(self.message)"
+    }
+}
+
+extension SwiftProtobufError: CustomDebugStringConvertible {
+    public var debugDescription: String {
+        "\(String(reflecting: self.code)): \(String(reflecting: self.message))"
+    }
+}
+
+// - MARK: Common errors
+
+extension SwiftProtobufError {
+    /// Errors arising from encoding protobufs into binary data.
+    public enum BinaryEncoding {
+        /// The definition of the message or one of its nested messages has required
+        /// fields but the message being encoded did not include values for them. You
+        /// must pass `partial: true` during encoding if you wish to explicitly ignore
+        /// missing required fields.
+        public static let missingRequiredFields = SwiftProtobufError(
```

Comment:
If we wanted to add more info in the future (say what fields were missing), how would we do that?  Would adding a `static func` as the same name but with args be safe?